### ISSUE TITLE
Regressions popup applet

### ISSUE THREAD

**Issue by wiiznokes** [pop-os/libcosmic#717]
### Before 0.13
![screenshot-2024-11-17-00-04-26](https://github.com/user-attachments/assets/f078a8df-02ed-4970-a88d-3d7b80c7610f)
![screenshot-2024-11-17-00-04-17](https://github.com/user-attachments/assets/7156687a-20a0-4081-ad59-00a36ed7ba80)

### After 0.13
![screenshot-2024-11-17-00-03-58](https://github.com/user-attachments/assets/5172a861-7136-4b4e-a228-7115c9865a3e)
![screenshot-2024-11-17-00-03-48](https://github.com/user-attachments/assets/faa3e0f2-e060-4676-8e7c-7ff684e94daa)

code:

```rs
if self.config.horizontal {
    get_layer_surface(SctkLayerSurfaceSettings {
        id: new_id,
        keyboard_interactivity: KeyboardInteractivity::OnDemand,
        anchor: layer_surface::Anchor::BOTTOM
            | layer_surface::Anchor::LEFT
            | layer_surface::Anchor::RIGHT,
        namespace: "clipboard manager".into(),
        size: Some((None, Some(350))),
        size_limits: Limits::NONE.min_width(1.0).min_height(1.0),
        ..Default::default()
    })
} else {
    let mut popup_settings =
        self.core
            .applet
            .get_popup_settings(Id::RESERVED, new_id, None, None, None);

    popup_settings.positioner.size_limits = Limits::NONE
        .max_width(400.0)
        .min_width(300.0)
        .min_height(200.0)
        .max_height(500.0);
    get_popup(popup_settings)
}
```

**Comment by Phosphorus-M** [comment]
I'm having the same problem :( 

With popups and surface the width is limited

**Comment by Phosphorus-M** [comment]
Technically, it could have a solution because this applet does something similar but without using a surface.

https://github.com/pop-os/cosmic-applets/tree/master/cosmic-applet-workspaces

But I'm too lazy to understand the way it's implemented

<img width="1919" height="742" alt="Image" src="https://github.com/user-attachments/assets/aabe9ad8-b9de-47d2-b1a1-fdd8f34a6eb3" />

**Comment by Lcstyle** [comment]
## Root cause identified

The regression is caused by iced commit [`cfe5f4b1`](https://github.com/pop-os/iced/commit/cfe5f4b1a4413b9c94c10832093b2a9e0de5eece) ("fix: skip resize if size is unchanged"), authored Oct 23, 2025.

This commit added early-return size checks to `winit_window.rs:request_surface_size()` for layer surfaces. The intent was to prevent redundant resize cycles, but it has a side effect: when the autosize widget measures content and the resulting size matches the surface's initial size, `resize_enabled` is never set to `true` for the window, causing `program.rs` to skip all rendering for that surface.

### How it manifests

The layer surface IS created (it captures keyboard input via `KeyboardInteractivity::Exclusive`), and `view_window()` IS called. But no buffer is ever committed to the Wayland compositor — the surface is invisible.

**Surfaces that trigger autosize to request a different size** (e.g., initial size 1200px but content only needs 360px) → `resize_enabled = true` → renders → visible.

**Surfaces where content fits the initial size** (or where autosize doesn't request a resize before the first frame) → `resize_enabled` stays `false` → `program.rs` skips the window → invisible forever.

### Proposed fix

In `iced/winit/src/platform_specific/wayland/sctk_event.rs`, the `LayerSurfaceEventVariant::Configure` handler (around line 878) should set `resize_enabled = true` on the first configure, ensuring the first frame always renders:

```rust
LayerSurfaceEventVariant::Configure(
    configure,
    surface,
    first,
) => {
    if let Some((id, w)) =
        surface_ids.get(&surface.id()).and_then(|id| {
            window_manager
                .get_mut(id.inner())
                .map(|v| (id.inner(), v))
        })
    {
        // FIX: Ensure the first configure always enables rendering.
        // Without this, if autosize doesn't request a resize (content
        // fits the initial size), resize_enabled stays false and
        // program.rs skips all rendering for this window.
        if first {
            w.resize_enabled = true;
        }

        let scale = w.state.scale_factor();
        // ... rest of existing code unchanged
```

This is a one-line addition that ensures layer surfaces always get at least one render pass after creation, without undoing the resize-skip optimization for subsequent configures.

### Workaround for applet developers

> @Phosphorus-M — Regarding the workspaces applet approach:

The workspaces applet avoids the issue by never using `get_layer_surface()` — it renders everything through `autosize` in the main `view()`. But that approach is for inline panel content, not popups.

For applets that need popups, the workaround is to use `get_popup()` (XDG popups) instead of `get_layer_surface()`. XDG popups go through a different code path in iced and are not affected by this bug. All official COSMIC applets (audio, battery, bluetooth, network, notifications, power) exclusively use `get_popup()` — none use `get_layer_surface()`.

@wash2

**Comment by Lcstyle** [comment]
1. Surface created → initial common.size = requested size (e.g., 360×530)
  2. First configure arrives → compositor says "your size is 360×530"
  3. Autosize widget says "I want 360×530" → request_surface_size() checks: guard.size == logical_size → early return (thanks to cfe5f4b1)
  4. Because no resize happened, resize_enabled stays false
  5. program.rs sees resize_enabled=false + no pending events → skips the window
  6. No frame is ever drawn → autosize never gets a chance to re-measure → goto step 3

A chicken-and-egg problem, a deadlock of sorts, because it has the same structure — two things each waiting on the other:

  - Rendering waits for resize_enabled=true
  - resize_enabled=true waits for a size change
  - A size change requires rendering (so autosize can measure content)

**Comment by Lcstyle** [comment]
## Update: PR submitted + additional compositor finding

### Fix PR

The proposed fix has been submitted as [pop-os/iced#258](https://github.com/pop-os/iced/pull/258) — a one-line change that sets `resize_enabled = true` on the first `LayerSurfaceEventVariant::Configure`.

### Additional finding: cosmic-comp does not send configure for partially-anchored layer surfaces

While testing the fix, I discovered a separate issue in **cosmic-comp**: layer surfaces that are not anchored to all four edges never receive a Wayland `configure` event from the compositor. This violates the [wlr-layer-shell protocol](https://wayland.app/protocols/wlr-layer-shell-unstable-v1), which states:

> "After creating a layer_surface object and setting it up, the client must perform an initial commit without any buffer attached. **The compositor will reply with a layer_surface.configure event.**"

Tested anchor configurations:

| Anchors | Layer | Configure received? |
|---------|-------|-------------------|
| TOP \| BOTTOM \| LEFT \| RIGHT | Overlay | Yes |
| TOP \| BOTTOM \| LEFT \| RIGHT | Top | Yes (via cursor-capture overlay) |
| TOP \| LEFT | Top | **No** |
| (empty) | Overlay | **No** |
| (empty) | Top | **No** |

The initial `wl_surface.commit()` is sent in all cases (confirmed in iced's `state.rs:get_layer_surface()`), but the compositor only responds with configure when all four edges are anchored.

This means applets cannot create centered or corner-positioned layer surfaces — only fullscreen ones. The workaround is to use all-edge anchoring with a fullscreen surface and center the popup content within it, bypassing `popup_container()`'s `Autosize` widget (which otherwise shrinks the surface to content size).

This may warrant a separate cosmic-comp issue.

**Comment by Lcstyle** [comment]
## Workaround: cursor-positioned popups via single-surface approach

While investigating this issue for the [clipboard-manager](https://github.com/cosmic-utils/clipboard-manager) applet, we developed a workaround for D-Bus triggered popups (keyboard shortcut → layer surface) that avoids the configure bug entirely.

### The problem

cosmic-comp only sends `configure` events for layer surfaces with all 4 edges anchored (`TOP|BOTTOM|LEFT|RIGHT`). Partially-anchored surfaces (e.g. `TOP|LEFT` with margin offsets for positioning) never receive configure, violating the [wlr-layer-shell spec](https://wayland.app/protocols/wlr-layer-shell-unstable-v1#zwlr_layer_surface_v1:event:configure):

> After creating a layer_surface object and setting it up, the client must perform an initial commit without any buffer attached. **The compositor will reply with a layer_surface.configure event.**

### The workaround: single-surface transition

Instead of creating a positioned layer surface (which never gets configure), we:

1. Create a **fullscreen** `Layer::Overlay` with all 4 edges anchored (this *does* get configure)
2. Use the surface first as a transparent `MouseArea` with `on_move` to capture cursor position
3. **Transition the same surface** from capture mode to popup mode — render the popup content at the captured cursor coordinates using container padding within the fullscreen overlay
4. Click outside the popup content dismisses the overlay

This avoids any destroy/create cycle and works entirely within a single layer surface that the compositor correctly configures.

### Commits

- [`fb67a8f`](https://github.com/Lcstyle/clipboard-manager/commit/fb67a8f) — Centered fullscreen overlay for D-Bus triggered popups (initial working baseline)
- [`11ecbf4`](https://github.com/Lcstyle/clipboard-manager/commit/11ecbf4) — Position popups at mouse cursor via single-surface transition

Key insight: `popup_container()` in libcosmic wraps content in an `Autosize` widget with `min_width=360, max_width=360, max_height=1000`, which controls the layer surface size. For fullscreen overlays, this must be bypassed — render styled content directly with custom containers instead.

**Comment by Lcstyle** [comment]
My comments aren't spam, I've been using the fix in my integrated features branch for at least a week so I know it resolves the issues.

https://github.com/cosmic-utils/clipboard-manager/compare/master...Lcstyle:clipboard-manager:integrated-features
